grant_disbursement/src/main.leo

```diff
@@ -0,0 +1,248 @@
+
+    assert(fully_vested_block > cliff_block);
+
+    return finalize_create_grant(
+      f0,
+      f1,
+      id,
+      credits_amount,
+      paleo_amount,
+      recipient_rewards_key,
+      recipient_principal_key,
+      cliff_block,
+      fully_vested_block
+      );
+  }
+
+  async function finalize_create_grant(
+    f0: Future,
+    f1: Future,
+    id: u64,
+    credits_amount: u64,
+    paleo_amount: u64,
+    recipient_rewards_key: address,
+    recipient_principal_key: address,
+    cliff_block: u32,
+    fully_vested_block: u32
+  ) {
+    f0.await();
+    f1.await();
+
+    let grant: Grant = Grant {
+      credits_amount: credits_amount,
+      paleo_amount: paleo_amount,
+      recipient_rewards_key: recipient_rewards_key,
+      recipient_principal_key: recipient_principal_key,
+      start_block: block.height,
+      cliff_block: cliff_block,
+      fully_vested_block: fully_vested_block
+    };
+
+    // Ensure the cliff block is in the future
+    assert(block.height < grant.cliff_block);
+
+    // Ensure the mapping doesn't already contain the id
+    assert(!grants.contains(id));
+
+    // Add the grant to the mapping
+    grants.set(id, grant);
+  }
+
+  async transition withdrawal_rewards(
+    id: u64,
+    paleo_amount: u64
+  ) -> Future {
+    // Transfer the pAleo to the recipient_rewards_key
+    let f0: Future = multi_token_support_program.aleo/transfer_public(PALEO_TOKEN_ID, self.caller, paleo_amount as u128);
+
+    return finalize_withdrawal_rewards(f0, id, paleo_amount, self.caller);
+  }
+
+  async function finalize_withdrawal_rewards(
+    f0: Future,
+    id: u64,
+    paleo_amount: u64,
+    caller: address
+  ) {
+    f0.await();
+
+    // Get the grant
+    let grant: Grant = grants.get(id);
+
+    // Ensure the caller is the recipient_rewards_key
+    assert_eq(caller, grant.recipient_rewards_key);
+
+    // Assert that the grant is before the cliff
+    assert(block.height < grant.cliff_block);
+
+    // Subtract the amount from the grant
+    let remaining_paleo: u64 = grant.paleo_amount - paleo_amount;
+
+    // Assert Pondo not in rebalancing state
+    let current_state: u8 = pondo_core_protocol.aleo/protocol_state.get(PROTOCOL_STATE_KEY);
+    assert(current_state != REBALANCING_STATE);
+
+    // Ensure the number of credits in the principal remains above the initial deposit
+    // Note: the paleo pool and the credits pool are always an underestimate (assuming no slashing) as it does not include rewards & owed commissions
+    // We use this underestimate because it as significantly reduces gas costs and should be close to the actual amount
+    // As it is an underestimate, the downside is it will slightly reduce the amount of paleo that can be withdrawn
+    let total_paleo_pool: u128 = multi_token_support_program.aleo/registered_tokens.get(PALEO_TOKEN_ID).supply;
+    let total_credits_pool: u64 = pondo_core_protocol.aleo/balances.get(DELEGATED_BALANCE);
```

Comment:
I want to call out the underestimate here. It will save a lot of gas but potentially reduce the number of pAleo a grantee can withdrawal until the next rebalance/deposit/withdrawal.